### DISCUSSION TITLE
Confused by the general flow that capnproto expects

### DISCUSSION THREAD

**Discussion by chbug** [capnproto/capnproto-rust#486] [Q&A]
Hey,

I'm investigating using capnproto serialization as a replacement for protobufs. I've looked at the examples, but I'm confused by what the API is trying to push me to do. I see the point of the various Builders, to set up an object that has the appropriate memory layout. But then in my use-case I'd expect to pass around the structured data in read-only format, and only eventually send it / store it. So what do people do in similar cases? Turn the Builder into a reader? Is this equivalent to serializing and passing it to a reader? The types are very bulky and none of the examples seem to show the life of an object besides immediate serialization/deserialization...

Thanks for your help!

**Comment by dwrensha** [comment]
Have you tried passing around a `message::Builder` or `message::TypedBuilder`? Those have `get_root_as_reader()` methods that allow read-only access.

**Comment by chbug** [comment]
I think one thing that confused me is that I expected `.into_reader()` to be the standard way of "closing" a builder, but actually you'd keep the builder around as the "live" object, right?

Unfortunately I have many more questions :) Let me know if this is not the right place to ask the followups...

- I assumed the favored approach was to use the `Typed` variants of Builder and Reader but I notice that `serialize::write_message` takes a plain `Builder`, and even the addressbook example use the plain variants. What's the recommended API?
- What's the difference between `message.init_root()` and `message.get_root()` when I'm first filling in an object? For Lists, initializing the size seems required, but for an object it seems that both work?

Thanks for your patience...

**Comment by dwrensha** [comment]
> I assumed the favored approach was to use the Typed variants of Builder and Reader but I notice that serialize::write_message takes a plain Builder, and even the addressbook example use the plain variants. What's the recommended API?

Maybe we should generalize the `serialize::write_message()` function so that it can accept any of `message::Builder`, `message::TypedBuilder`, `message::Reader`, or `message::TypedReader`. It looks like the `message::ReaderSegments` trait is all that `write_message()` actually needs, and `message::Builder<A>` already implements that.

> What's the difference between message.init_root() and message.get_root() when I'm first filling in an object? For Lists, initializing the size seems required, but for an object it seems that both work?

There is some explanation of `get` vs `init` in the C++ docs: https://capnproto.org/cxx.html#structs

The main thing is that if there is an existing value in the field, `get` retains that value, while `init` clobbers it.

**Comment by chbug** [comment]
Thanks for the `get` vs `init` link, very useful.

What are the typed variants solving? Granted I only wrote toy code, but I don't see a big difference compared to the plain builder/reader. So if it's not buying much and is not fully integrated, I'm happy to stay clear, but I'm curious what the intent is.

(and after that I think I'll be good for a little while :P)

**Reply by chbug** [reply]
Ok, I could have tried a bit harder. So the typed variants are just here to help avoid passing around completely typeless `message::Builder`. In which case, indeed generalizing the various serializers to accept them would make sense, as I think one needs to do an into_inner() to do most things, which consumes the typed struct :)

**Comment by andrewbanchich** [comment]
@dwrensha i'm assessing capnproto for a project and ran into this, too. it would be great to have:

1. doc comments on the generated code. i can't figure out from looking at the fn signatures what i'm supposed to do.
2. some rust examples, ideally in a doc site, but also in the doc comments, instead of relying on the rust impl being the same as the C++ one and having rust programmers need to read languages they're unfamiliar with.

is this something that's planned?
